source/backends/ubuntu/ubupkg.d

```diff
@@ -0,0 +1,167 @@
+        }
+    }
+
+private:
+    string globalTmpDir;
+    string langpackDir;
+    string localeDir;
+    string[] langpackLocales;
+    Array!Package allPackages;
+
+    void extractLangpacks ()
+    {
+        import std.algorithm : filter, map;
+        import std.array : appender, array, split;
+        import std.file : dirEntries, exists, SpanMode, readText;
+        import std.parallelism : parallel;
+        import std.path : baseName;
+        import std.process : Pid, spawnProcess, wait;
+        import std.string : splitLines, startsWith;
+
+        auto path = buildPath (langpackDir, "usr", "share", "locale-langpack");
+
+        if (!langpackDir.exists) {
+            foreach (pkg; allPackages) {
+                if (!pkg.name.startsWith ("language-pack"))
+                    continue;
+
+                /* :( */
+                UbuntuPackage upkg = cast (UbuntuPackage) pkg;
+
+                logDebug ("Extracting %s", pkg.name);
+                upkg.extractPackage (langpackDir);
+            }
+
+            auto supportedd = buildPath (langpackDir, "var", "lib", "locales", "supported.d");
+
+            mkdirRecurse (localeDir);
+
+            foreach (locale; parallel (supportedd.dirEntries (SpanMode.shallow), 5))
+            {
+                    foreach (ref line; locale.readText.splitLines) {
+                            auto components = line.split (" ");
+                            auto localecharset = components[0].split (".");
+
+                            auto outdir = buildPath (localeDir, components[0]);
+                            logDebug ("Generating locale in %s", outdir);
+
+                            auto pid = spawnProcess (["/usr/bin/localedef",
+                                                      "--no-archive",
+                                                      "-i",
+                                                      localecharset[0],
+                                                      "-c",
+                                                      "-f",
+                                                      components[1],
+                                                      outdir]);
+
+                            scope (exit) wait (pid);
+                    }
+            }
+
+        }
+
+        if (langpackLocales is null)
+            langpackLocales = localeDir.dirEntries (SpanMode.shallow)
+                .filter!(f => f.isDir)
+                .map!(f => f.name.baseName)
+                .array;
+    }
+
+    string[string] getTranslations (const string domain, const string text)
+    {
+        import core.stdc.locale;
+
+        import std.c.stdlib : getenv, setenv, unsetenv;
+        import std.string : fromStringz, toStringz;
+
+        foreach (ref var; ["LC_ALL", "LANG", "LANGUAGE", "LC_MESSAGES"])
+                unsetenv (var.toStringz);
+
+        setenv ("LOCPATH", localeDir.toStringz, true);
+
+        auto initialLocale = setlocale (LC_ALL, "");
```

Comment:
On Tue, Aug 30, 2016 at 08:23:41AM -0700, Matthias Klumpp wrote:
> The slightly annoying result of that will be that we'll need to package libmo in Debian and Ubuntu first, or abandon the dub-based build system in favor of Meson, because dub is unable to check for the existence of libraries and is also unable to conditionally compile stuff based on flags (otherwise one could just disable the Ubuntu backend if they don't have libmo).
> 
> On the other hand, libmo will likely be super-low on maintenance, so getting it into Debian will be easy.

I would kill dub if it can't do what we want any more. Would ideally use
libmo as a submodule for a while until we're sure it is good enough to
stand alone.

-- 
Iain Lane                                  [ iain@orangesquash.org.uk ]
Debian Developer                                   [ laney@debian.org ]
Ubuntu Developer                                   [ laney@ubuntu.com ]